### PR TITLE
W2v2

### PR DESCRIPTION
**Pull request by taylanbil**

### CODE REVIEW

File: fairseq/criterions/wav2vec_criterion.py [hunk 1]
```diff
@@ -53,11 +63,24 @@ def forward(self, model, sample, reduce=True, log_pred=False):
...
+            loss = F.cross_entropy(
+                logits, target, reduction="sum" if reduce else "none",
+                ignore_index=-1,
+            )
         else:
-            loss = F.binary_cross_entropy_with_logits(logits, target.float(), weights, reduction="sum" if reduce else "none",)
+            loss = F.binary_cross_entropy_with_logits(
+                logits, target.float(), weights,
+                reduction="sum" if reduce else "none",
+                ignore_index=-1,
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py]: this wont work, binary_cross_entropy doesnt have ignore_index. in this case you should compute unreduced loss and then fill unmasked parts with 0s (if mask is available)

File: fairseq/criterions/wav2vec_criterion.py [hunk 2]
```diff
@@ -53,11 +63,24 @@ def forward(self, model, sample, reduce=True, log_pred=False):
...
+            )
 
-        sample_size = target.numel() if self.infonce else target.long().sum().item()
+        if 'sample_size' in sample and self.infonce:
+            sample_size = sample['sample_size']
+        elif 'mask_indices' in sample['net_input'] and self.infonce:
+            # XXX: what happens if not self.infonce?
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py]: should be the same thing, you can just remove the "self.infonce" check

File: fairseq/criterions/wav2vec_criterion.py [hunk 3]
```diff
@@ -99,13 +126,14 @@ def forward(self, model, sample, reduce=True, log_pred=False):
                     max = logits.argmax(-1) == 0
                     min = logits.argmin(-1) == 0
                     both = max & min
-                    corr = max.long().sum().item() - both.long().sum().item()
-                    count = max.numel()
+                    # corr = max.long().sum().item() - both.long().sum().item()
+                    corr = max.long().sum() - both.long().sum()
+                    count = float(max.numel())
```
Comment (alexeib) [inline fairseq/criterions/wav2vec_criterion.py]: you need to make sure that unmasked states dont contribute to either corr or count here

File: fairseq/modules/gumbel_vector_quantizer.py [hunk 4]
```diff
@@ -160,6 +160,7 @@ def forward(self, x, produce_targets=False):
         avg_probs = torch.softmax(
             x.view(bsz * tsz, self.groups, -1).float(), dim=-1
         ).mean(dim=0)
+        avg_probs = avg_probs.detach()
```
Comment (alexeib) [inline fairseq/modules/gumbel_vector_quantizer.py]: i think this might break the diversity penalty (it is applied on prob perplexity below that uses avg_probs)

**Review by alexeib — COMMENTED** [review]
cant see anything that is obviously wrong, though see some comments inline (in particular, diversity penalty is prob broken, but that wouldnt be root cause of what we see).

one issue is that you are sampling negatives also from padded states now. this is something you are addressing in next pr but that one has issues (see comments there)

have you checked that the generated masks are different each time and have sufficient number of masked spans, etc?

File: fairseq/criterions/wav2vec_criterion.py (hunk 1 above)
Comment (taylanbil) [reply fairseq/criterions/wav2vec_criterion.py]: I'll change the whole computation to be

1. compute unreduced loss
2. loss = loss * mask
3. loss = loss.sum()

File: fairseq/criterions/wav2vec_criterion.py (hunk 2 above)
Comment (taylanbil) [reply fairseq/criterions/wav2vec_criterion.py]: sg

File: fairseq/criterions/wav2vec_criterion.py (hunk 3 above)
Comment (taylanbil) [reply fairseq/criterions/wav2vec_criterion.py]: yep, will roll the changes from other pr to this.

File: fairseq/modules/gumbel_vector_quantizer.py (hunk 4 above)
Comment (taylanbil) [reply fairseq/modules/gumbel_vector_quantizer.py]: ah.. I'll remove.

**Review by taylanbil — COMMENTED** [review]
> one issue is that you are sampling negatives also from padded states now. this is something you are addressing in next pr but that one has issues (see comments there)

When I find time I intend to clean this pr and roll some changes from the other one into this.

> have you checked that the generated masks are different each time and have sufficient number of masked spans, etc?

no, I have not. how do I check that?

**Comment by alexeib** [comment]
just print out the masks and/or statistics about the masks from the collater?

**Comment by taylanbil** [comment]
Addressed comments, pushed.

What stats are you interested in?
